src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -390,10 +446,10 @@ private Task ProcessEventAsync(CompilationEvent e, CancellationToken cancellatio
 
         private async Task ProcessSymbolDeclaredAsync(SymbolDeclaredCompilationEvent symbolEvent, CancellationToken cancellationToken)
         {
-            // Create a task per-analyzer to execute analyzer actions.
-            // We execute analyzers in parallel, but for a given analyzer we execute actions sequentially.
-            var tasksMap = PooledDictionary<DiagnosticAnalyzer, Task>.GetInstance();
-
+            // Collect all the analyzer action executors grouped by analyzer.
+            // NOTE: Right now we execute all the actions sequentially, but there is scope to fine tune this to execute certain actions in parallel.
+            var actionsMap = PooledDictionary<DiagnosticAnalyzer, ArrayBuilder<Action>>.GetInstance();
```

Comment:
Instead of building a big action map and then running them, I recommend getting rid of the action map, adding Task.Run around this entire method body, and just doing the actions directly.